ussologin/ussologin.go

```diff
@@ -83,38 +89,35 @@ var loginForm = form.Form{
 	},
 }
 
-// LoginWithToken completes a login attempt using tok. The ussoAuthURL
-// should have been obtained from the UbuntuSSOOAuth field in a response
-// to a LoginMethods request from the target service.
-func LoginWithToken(client *http.Client, ussoAuthUrl string, tok *usso.SSOData) error {
-	req, err := http.NewRequest("GET", ussoAuthUrl, nil)
-	if err != nil {
-		return errgo.Notef(err, "cannot create request")
-	}
-	base := *req.URL
-	base.RawQuery = ""
-	rp := usso.RequestParameters{
-		HTTPMethod:      req.Method,
-		BaseURL:         base.String(),
-		Params:          req.URL.Query(),
-		SignatureMethod: usso.HMACSHA1{},
-	}
-	if err := tok.SignRequest(&rp, req); err != nil {
-		return errgo.Notef(err, "cannot sign request")
-	}
-	resp, err := client.Do(req)
-	if err != nil {
-		return errgo.Notef(err, "cannot do request")
+type TokenGetter interface {
+	GetToken(context.Context) (*usso.SSOData, error)
+}
+
+type StoreTokenGetter struct {
+	Store       TokenStore
+	TokenGetter TokenGetter
+}
+
+// GetToken implements TokenGetter.GetToken. If a store is specified then GetToken will try to retrieve a token from there.
+func (g StoreTokenGetter) GetToken(ctx context.Context) (*usso.SSOData, error) {
+	err := errgo.New("store not configured")
+	if g.Store != nil {
+		var tok *usso.SSOData
+		tok, err = g.Store.Get()
+		if err == nil {
+			return tok, nil
+		}
 	}
-	defer resp.Body.Close()
-	if resp.StatusCode == http.StatusOK {
-		return nil
+	if g.TokenGetter == nil {
```

Comment:
When would it be valid to have a StoreTokenGetter with no underlying TokenGetter ?

Reply:
When you just want to use a stored version if it's there, but fail if its not.